src/eventbridge/eventpattern/index.ts

```diff
@@ -0,0 +1,798 @@
+      const parent = getEventReference(expression.expr);
+      if (parent) {
+        return [...parent, key];
+      }
+      return undefined;
+    }
+    return undefined;
+  };
+
+  // find the return, we'll the resolve the rest as needed.
+  const [ret] = predicate.body.statements.filter((expr) => isReturn(expr));
+
+  if (!ret) {
+    throw new Error("Missing return statement in predicate block.");
+  }
+
+  const result = handleSmnt(ret);
+
+  return patternDocumentToEventPattern(result);
+};
+
+type EventReference = [string] | ["detail", ...string[]];
+
+// TODO: validate again object schema?
+function assertValidEventRefererence(
+  eventReference: string[]
+): asserts eventReference is EventReference {
+  if (eventReference.length === 0) {
+    throw new Error("Direct use of the event is invalid.");
+  }
+  if (eventReference.length > 1) {
+    const [first] = eventReference;
+    if (first !== "detail") {
+      throw `Event references with depth greater than one must be on the detail propert, got ${eventReference.join(
+        ","
+      )}`;
+    }
+  }
+}
+
+/**
+ * Retrieves a string, number, boolean, undefined, or null constant from the given expression.
+ * Wrap the value to not be ambiguous with the undefined value.
+ * When one is not found, return undefined (not wrapped).
+ *
+ * TODO: Support following references.
+ */
+const getConstant = (
+  expr: Expr
+): { value: string | number | boolean | undefined | null } | undefined => {
+  if (
+    isStringLiteralExpr(expr) ||
+    isNumberLiteralExpr(expr) ||
+    isBooleanLiteral(expr)
+  ) {
+    return { value: expr.value };
+  } else if (isNullLiteralExpr(expr)) {
+    return { value: expr.undefined ? undefined : null };
+  } else if (isUnaryExpr(expr) && expr.op === "-") {
+    const number = assertNumber(getConstant(expr.expr)?.value);
+    return { value: -number };
+  }
+  return undefined;
+};
+
+/**
+ * Normalize retrieving the name of a property between a element access and property access expression.
+ */
+const getPropertyAccessKey = (
+  expr: PropAccessExpr | ElementAccessExpr
+): string => {
+  return isPropAccessExpr(expr)
+    ? expr.name
+    : assertString(getConstant(expr.element)?.value);
+};
+
+/**
+ * { isPresent: true } => { isPresent: false }
+ * { value: true } => { value: false }
+ * { value "a" } => { "anything-but": "a" }
+ * Do not support
+ *   * documents with more than one field in in - do not support OR accross multiple fields, inverting multiple AND fields is OR
+ *   * aggregate pattern (OR logic) - do not support AND within a field, inverting OR statements becomes AND, the one exception are numeric ranges.
+ */
+const negateDocument = (classDocument: PatternDocument): PatternDocument => {
+  if (Object.keys(classDocument).length > 1) {
+    throw Error(
+      "Can only negate simple statements like equals, doesn't equals, and prefix."
+    );
+  }
+  const [key = undefined] = Object.keys(classDocument.doc);
```

Comment:
what does this syntax do?
```ts
const [key = undefined]
```

Never seen that before.

Reply:
first element of the array, initialize to undefined if there is no first element.

```ts
const key = Object.keys(classDocument.doc).length > 0 ? Object.keys(classDocument.doc)[0] : undefined.
```

Reply:
Love that.